beep/tests/test_maccor_to_biologic_mb.py

```diff
@@ -159,34 +246,428 @@ def test_discharge_current_step_conversion(self):
+            "      <Ends/>"
+            "      <Reports/>"
+            "      <Range></Range>"
+            "      <Option1></Option1>"
+            "      <Option2></Option2>"
+            "      <Option3></Option3>"
+            "      <StepNote></StepNote>"
+            "    </TestStep>"
+            "    <TestStep>"
+            "      <StepType>  End   </StepType>"
+            "      <StepMode>        </StepMode>"
+            "      <StepValue></StepValue>"
+            "      <Limits/>"
+            "      <Ends/>"
+            "      <Reports/>"
+            "      <Range></Range>"
+            "      <Option1></Option1>"
+            "      <Option2></Option2>"
+            "      <Option3></Option3>"
+            "      <StepNote></StepNote>"
+            "    </TestStep>"
+            "  </ProcSteps>"
+            "</MaccorTestProcedure>"  
+        ))
+
+    
+
+        expected_output = (
+            "BT-LAB SETTING FILE\r\n"
+            "\r\n"
+            "Number of linked techniques : 1\r\n"
+            "\r\n"
+            "Filename : C:\dummy.mps\r\n"
+            "\r\n"
+            "Device : BCS-805\r\n"
+            "Ecell ctrl range : min = 0.00 V, max = 10.00 V\r\n"
+            "Electrode material : \r\n"
+            "Initial state : \r\n"
+            "Electrolyte : \r\n"
+            "Comments : \r\n"
+            "Mass of active material : 0.001 mg\r\n"
+            " at x = 0.000\r\n"
+            "Molecular weight of active material (at x = 0) : 0.001 g/mol\r\n"
+            "Atomic weight of intercalated ion : 0.001 g/mol\r\n"
+            "Acquisition started at : xo = 0.000\r\n"
+            "Number of e- transfered per intercalated ion : 1\r\n"
+            "for DX = 1, DQ = 26.802 mA.h\r\n"
+            "Battery capacity : 2.280 mA.h\r\n"
+            "Electrode surface area : 0.001 cm\N{superscript two}\r\n"
+            "Characteristic mass : 9.130 mg\r\n"
+            "Cycle Definition : Loop\r\n"
+            "Turn to OCV between techniques\r\n"
+            "\r\n"
+            "Technique : 1\r\n"
+            "Modulo Bat\r\n"
+            "Ns                  0                   1                   2                   3                   4                   5                   6                   7                   8                   \r\n"
+            "type                Rest                CC                  CV                  CV                  CV                  CC                  Loop                Loop                Loop                \r\n"
+            "Apply I/C           I                   C / N               C / N               C / N               C / N               C / N               C / N               C / N               C / N               \r\n"
+            "ctrl1_val                               1.000               3.300               3.300               3.300               142.900             100.000             100.000             100.000             \r\n"
```

Comment:
Some of these control values don't look quite right given the procedure values. Type `CC` should correspond to StepMode `Current` and Type `CV` should correspond to StepMode `Voltage` I think the Limit values from the maccor are a little tricky, but they should probably go into `ctrl2_val` and `ctrl2_val_unit`
Let me know if that makes sense?

Reply:
I will make the changes for `Limits` then.

`Current` corresponds to `CC` and `Voltage` to CV you say. This is the output I'm expecting, if that's unclear, let me break down the conversion:

1. Step 1 is `Rest` so seq 0 is `Rest`
2. Step 2 is `Current` so seq 1 is `CC`
3. Step 3 is `Do 1` which is not representable as a seq
4. Step 4 is `Voltage` so seq 2 is `CV`
5. Step 5 is `Loop 1` paired with step 3, it's value is 2 which means all steps between it and the opening do will occur 2 more times for a total of 3 occurrences. Since there is no advance cycle at the end of the loop, the loop must unroll. This means seq 3 and seq 4 are `CV`, copies of seq 2.
6. Step 6 is `Do 1` which is not representable as a seq
7. Step 7 is `Current` so seq 5 is `CC`
8. Step 8 is  `Adv Cycle`,  it fulfills the conditions for the containing loop not to unroll and so is not converted to a seq.
9. Step 9 is `Loop 1`, it does not unroll so seq 6 is `Loop` and loops to seq 5.
10. Step 10 is `Adv Cycle`, it is not in a loop, so the functionality is mocked by adding 2 loop steps as documented in the PR desciption above, these are seqs 7 and 8.